File: 05Feb2025/backend/backend.py
from fastapi import FastAPI
from fastapi.encoders import jsonable_encoder
from pydantic import BaseModel
from typing import Dict, List
import json
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

BOOKS_FILE = "../storage/books.json"
MEMBERS_FILE = "../storage/members.json"

origins = [
    "http://localhost",
    "http://localhost:8000",
    "http://localhost:3000",
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

class AllocationDetails(BaseModel):
    id: str
    book_name: str
    from_date: str
    to_date: str

class Book(BaseModel):
    isbn: str
    name: str
    author: str
    total_copies: int
    allocated_copies: int

class Member(BaseModel):
    id: int
    name: str
    allocated_books: List[AllocationDetails]
    
def loadData(fileName):
    with open(fileName, "r") as fileData:
        return json.load(fileData)

members = loadData(MEMBERS_FILE)
member_id = members["members"][-1]["id"] + 1 if members["members"][-1]["id"] else 0

def saveData(fileName, data):
    with open(fileName, "w") as file:
        json.dump(data, file, indent=4)

@app.get("/books")
def getBooks():
    books = loadData(BOOKS_FILE)
    return books["books"]

@app.get("/books/{book_isbn}")
def getBook(book_isbn:int):
    books = loadData(BOOKS_FILE)
    returnBook = {}
    for book in books["books"]:
        if(book["isbn"] == book_isbn):
            returnBook = book
    return returnBook

@app.post("/books")
def addBook(book: Book):
    books = loadData(BOOKS_FILE)
    jsonBookData = jsonable_encoder(book)
    books["books"].append(jsonBookData)
    saveData(BOOKS_FILE, books)
    return book

@app.put("/books/{book_isbn}")
def updateBook(book_isbn:str, bookData:Book):
    jsonBookData = jsonable_encoder(bookData)
    print(jsonBookData)
    books = loadData(BOOKS_FILE)
    for book in books["books"]:
        if(book["isbn"] == book_isbn):
            book["name"] = jsonBookData["name"]
            book["author"] = jsonBookData["author"]
            book["total_copies"] = jsonBookData["total_copies"]
            book["allocated_copies"] = jsonBookData["allocated_copies"]
    saveData(BOOKS_FILE, books)
    return "Success"

@app.delete("/books/{book_isbn}")
def deleteBook(book_isbn):
    books = loadData(BOOKS_FILE)
    bookToDelete = {}
    for book in books["books"]:
        if(book["isbn"] == book_isbn):
            bookToDelete = book
    books["books"].remove(bookToDelete)
    return "Success"

@app.get("/members")
def getMembers():
    return loadData(MEMBERS_FILE)

@app.get("/members/{member_id}")
def getMember(member_id:int):
    members = loadData(MEMBERS_FILE)
    returnMember = {}
    for member in members["members"]:
        if(member["id"] == member_id):
            returnMember = member
    return returnMember

@app.post("/members")
def addMember(member_data: Dict):
    global member_id
    members = loadData(MEMBERS_FILE)
    print(member_data)
    memberData = {
        "id": member_id,
        "name": member_data['member_name'],
        "allocated_books": []
    }
    jsonMemberData = jsonable_encoder(memberData)
    members["members"].append(jsonMemberData)
    member_id += 1
    saveData(MEMBERS_FILE, members)
    return memberData

@app.put("/members/{member_id}")
def updateMember(member_id:int, memberData:Member):
    members = loadData(MEMBERS_FILE)
    jsonMemberData = jsonable_encoder(memberData)
    for member in members["members"]:
        if(member["id"] == member_id):
            member["name"] = jsonMemberData["name"]
            member["allocated_books"] = jsonMemberData["allocated_books"]
    saveData(MEMBERS_FILE, members)
    return "Success"


@app.delete("/members/{member_id}")
def deleteMember(member_id):
    members = loadData(MEMBERS_FILE)
    memberToDelete = {}
    for member in members["members"]:
        if(member["id"] == member_id):
            memberToDelete = member
    members["members"].remove(memberToDelete)
    return "Success"

@app.put("/allocateBook/{book_isbn}/{to_member_id}")
def allocateBook(book_isbn: str, to_member_id: int, allocation_data: Dict):
    members = loadData(MEMBERS_FILE)
    books = loadData(BOOKS_FILE)
    for book in books["books"]:
        if book["isbn"] == book_isbn:
            if(book["allocated_copies"] >= book["total_copies"]):
                return {"error": "All copies of the book are already allocated"}
            else:
                book["allocated_copies"] += 1
    
    for member in members["members"]:
        if member["id"] == to_member_id:
            for allocated_book in member["allocated_books"]:
                if allocated_book["id"] == book_isbn:
                    return {"error": "Book already allocated to this member"}
            
            member["allocated_books"].append({
                "id": book_isbn,
                "book_name": allocation_data["book_name"],
                "from_date": allocation_data["from_date"],
                "to_date": allocation_data["to_date"]
            })

    saveData(MEMBERS_FILE, members)    
    saveData(BOOKS_FILE, books)

    return {"message": "Success"}


@app.put("/deallocateBook/{book_isbn}/{member_id}")
def deallocateBook(member_id: int, book_isbn: str):
    members = loadData(MEMBERS_FILE)
    books = loadData(BOOKS_FILE)
    
    for book in books["books"]:
        if book["isbn"] == book_isbn:
            book["allocated_copies"] -= 1
    saveData(BOOKS_FILE, books)

    for member in members["members"]:
        if member["id"] == member_id:
            # Find and remove the allocated book entry
            member["allocated_books"] = [
                book for book in member["allocated_books"] if book["id"] != book_isbn
            ]
    saveData(MEMBERS_FILE, members)

    return {"message": "Success"}

File: 05Feb2025/backend/backend.js
const express = require('express');
const fs = require('fs').promises;
const cors = require('cors');
const app = express();

// Middleware
app.use(express.json());
app.use(cors());

const BOOKS_FILE = "../storage/books.json";
const MEMBERS_FILE = "../storage/members.json";

// Helper functions
async function loadData(fileName) {
    try {
        const data = await fs.readFile(fileName, 'utf8');
        return JSON.parse(data);
    } catch (error) {
        if (error.code === 'ENOENT') {
            return fileName === BOOKS_FILE ? 
                { books: [] } : 
                { members: [] };
        }
        throw error;
    }
}

async function saveData(fileName, data) {
    try {
        const formattedData = fileName === BOOKS_FILE ? 
            { books: Array.isArray(data.books) ? data.books : [] } :
            { members: Array.isArray(data.members) ? data.members : [] };
            
        await fs.writeFile(fileName, JSON.stringify(formattedData, null, 4));
    } catch (error) {
        console.error(`Error writing to ${fileName}:`, error);
        throw error;
    }
}

// Book routes
app.get('/books', async (req, res) => {
    try {
        const data = await loadData(BOOKS_FILE);
        res.json(data.books);
    } catch (error) {
        res.status(500).json({ error: 'Failed to load books' });
    }
});

app.get('/books/:isbn', async (req, res) => {
    try {
        const data = await loadData(BOOKS_FILE);
        const book = data.books.find(book => book.isbn === req.params.isbn);
        if (!book) {
            return res.status(404).json({ error: 'Book not found' });
        }
        res.json(book);
    } catch (error) {
        res.status(500).json({ error: 'Failed to load book' });
    }
});

app.post('/books', async (req, res) => {
    try {
        const { isbn, name, author, total_copies, allocated_copies } = req.body;
        
        if (!isbn || !name || !author || total_copies === undefined) {
            return res.status(400).json({ error: 'Missing required fields' });
        }

        const data = await loadData(BOOKS_FILE);
        
        if (data.books.some(book => book.isbn === isbn)) {
            return res.status(409).json({ error: 'Book with this ISBN already exists' });
        }

        const newBook = {
            isbn,
            name,
            author,
            total_copies: parseInt(total_copies),
            allocated_copies: allocated_copies || 0
        };

        data.books.push(newBook);
        await saveData(BOOKS_FILE, data);
        res.status(201).json(newBook);
    } catch (error) {
        res.status(500).json({ error: 'Failed to add book' });
    }
});

app.put('/books/:isbn', async (req, res) => {
    try {
        const { name, author, total_copies, allocated_copies } = req.body;
        const data = await loadData(BOOKS_FILE);
        const bookIndex = data.books.findIndex(book => book.isbn === req.params.isbn);
        
        if (bookIndex === -1) {
            return res.status(404).json({ error: 'Book not found' });
        }

        data.books[bookIndex] = {
            ...data.books[bookIndex],
            name: name || data.books[bookIndex].name,
            author: author || data.books[bookIndex].author,
            total_copies: total_copies !== undefined ? parseInt(total_copies) : data.books[bookIndex].total_copies,
            allocated_copies: allocated_copies !== undefined ? parseInt(allocated_copies) : data.books[bookIndex].allocated_copies
        };

        await saveData(BOOKS_FILE, data);
        res.json(data.books[bookIndex]);
    } catch (error) {
        res.status(500).json({ error: 'Failed to update book' });
    }
});

app.delete('/books/:isbn', async (req, res) => {
    try {
        const data = await loadData(BOOKS_FILE);
        const bookIndex = data.books.findIndex(book => book.isbn === req.params.isbn);
        
        if (bookIndex === -1) {
            return res.status(404).json({ error: 'Book not found' });
        }

        data.books.splice(bookIndex, 1);
        await saveData(BOOKS_FILE, data);
        res.json({ message: 'Book deleted successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to delete book' });
    }
});

// Member routes
app.get('/members', async (req, res) => {
    try {
        const data = await loadData(MEMBERS_FILE);
        res.json(data);
    } catch (error) {
        res.status(500).json({ error: 'Failed to load members' });
    }
});

app.get('/members/:id', async (req, res) => {
    try {
        const data = await loadData(MEMBERS_FILE);
        const member = data.members.find(member => member.id === parseInt(req.params.id));
        if (!member) {
            return res.status(404).json({ error: 'Member not found' });
        }
        res.json(member);
    } catch (error) {
        res.status(500).json({ error: 'Failed to load member' });
    }
});

app.post('/member', async (req, res) => {
    try {
        const { name } = req.body;
        
        if (!name) {
            return res.status(400).json({ error: 'Name is required' });
        }

        const data = await loadData(MEMBERS_FILE);
        
        // Generate new ID
        const maxId = data.members.reduce((max, member) => Math.max(max, member.id), 0);
        const newMember = {
            id: maxId + 1,
            name,
            allocated_books: []
        };

        data.members.push(newMember);
        await saveData(MEMBERS_FILE, data);
        res.status(201).json(newMember);
    } catch (error) {
        res.status(500).json({ error: 'Failed to add member' });
    }
});

app.put('/members/:id', async (req, res) => {
    try {
        const { name, allocated_books } = req.body;
        const data = await loadData(MEMBERS_FILE);
        const memberIndex = data.members.findIndex(member => member.id === parseInt(req.params.id));
        
        if (memberIndex === -1) {
            return res.status(404).json({ error: 'Member not found' });
        }

        data.members[memberIndex] = {
            ...data.members[memberIndex],
            name: name || data.members[memberIndex].name,
            allocated_books: allocated_books || data.members[memberIndex].allocated_books
        };

        await saveData(MEMBERS_FILE, data);
        res.json(data.members[memberIndex]);
    } catch (error) {
        res.status(500).json({ error: 'Failed to update member' });
    }
});

app.delete('/members/:id', async (req, res) => {
    try {
        const data = await loadData(MEMBERS_FILE);
        const memberIndex = data.members.findIndex(member => member.id === parseInt(req.params.id));
        
        if (memberIndex === -1) {
            return res.status(404).json({ error: 'Member not found' });
        }

        data.members.splice(memberIndex, 1);
        await saveData(MEMBERS_FILE, data);
        res.json({ message: 'Member deleted successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to delete member' });
    }
});

// Book allocation routes
app.put('/allocateBook/:isbn/:memberId', async (req, res) => {
    try {
        const { isbn, memberId } = req.params;
        const { from_date, to_date } = req.body;
        const membersData = await loadData(MEMBERS_FILE);
        const booksData = await loadData(BOOKS_FILE);

        const member = membersData.members.find(m => m.id === parseInt(memberId));
        if (!member) {
            return res.status(404).json({ error: 'Member not found' });
        }

        const book = booksData.books.find(b => b.isbn === isbn);
        if (!book) {
            return res.status(404).json({ error: 'Book not found' });
        }

        if (book.allocated_copies >= book.total_copies) {
            return res.status(400).json({ error: 'No copies available for allocation' });
        }

        // Check if the book is already allocated to this member
        const existingAllocation = member.allocated_books.find(entry => entry.id === isbn);
        if (existingAllocation) {
            return res.status(400).json({ error: 'Book already allocated to this member' });
        }

        // Allocate book
        member.allocated_books.push({ id: isbn, from_date:from_date, to_date:to_date });
        book.allocated_copies += 1;

        await saveData(MEMBERS_FILE, membersData);
        await saveData(BOOKS_FILE, booksData);
        res.json({ message: 'Book allocated successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to allocate book' });
    }
});


app.put('/deallocateBook/:isbn/:memberId', async (req, res) => {
    try {
        const { isbn, memberId } = req.params;
        const membersData = await loadData(MEMBERS_FILE);
        const booksData = await loadData(BOOKS_FILE);

        const member = membersData.members.find(m => m.id === parseInt(memberId));
        if (!member) {
            return res.status(404).json({ error: 'Member not found' });
        }

        const book = booksData.books.find(b => b.isbn === isbn);
        if (!book) {
            return res.status(404).json({ error: 'Book not found' });
        }

        // Remove the book from allocated list
        const prevAllocatedLength = member.allocated_books.length;
        member.allocated_books = member.allocated_books.filter(entry => entry.id !== isbn);

        if (member.allocated_books.length === prevAllocatedLength) {
            return res.status(400).json({ error: 'Book not allocated to this member' });
        }

        // Reduce allocated copies count
        book.allocated_copies = Math.max(0, book.allocated_copies - 1);

        await saveData(MEMBERS_FILE, membersData);
        await saveData(BOOKS_FILE, booksData);
        res.json({ message: 'Book deallocated successfully' });
    } catch (error) {
        res.status(500).json({ error: 'Failed to deallocate book' });
    }
});

const PORT = 8000;
app.listen(PORT, () => {
    console.log(`Server is running on port ${PORT}`);
});
